cloudigrade/account/reports/helper.py

```diff
@@ -0,0 +1,115 @@
+"""Cloud-specific helpers to aid in cloud-agnostic report processing."""
+from abc import ABC, abstractmethod
+
+from django.db import models
+from django.utils.translation import gettext as _
+
+from account import CLOUD_PROVIDERS
+from account.exceptions import InvalidCloudProviderError
+from account.models import AwsAccount
+
+
+def get_report_helper(cloud_provider, cloud_account_id):
+    """
+    Get a cloud-specific ReportHelper instance.
+
+    Args:
+        cloud_provider (str): The cloud provider
+        cloud_account_id (str): The cloud-specific account ID
+
+    Returns:
+        ReportHelper: An instance of a cloud-specific ReportHelper child class.
+
+    """
+    if cloud_provider not in CLOUD_PROVIDERS:
+        raise InvalidCloudProviderError(
+            _('Unsupported cloud provider "{0}".').format(cloud_provider)
+        )
+
+    if cloud_provider == 'aws':
+        return AwsReportHelper(cloud_account_id)
+
+    raise NotImplementedError()
+
+
+class ReportHelper(ABC):
+    """Abstract base class for cloud-specific reporting helper functions."""
+
+    def __init__(self, cloud_account_id):
+        """
+        Initialize a new ReportHelper instance.
+
+        Args:
+            cloud_account_id (str): The cloud-specific account ID
+        """
+        self.cloud_account_id = cloud_account_id
+
+    @abstractmethod
+    def assert_account_exists(self):
+        """Assert that the cloud-specific account ID exists in cloudigrade."""
+
+    @property
+    @abstractmethod
+    def instance_account_filter(self):
+        """Get a Django query filter to restrict instances to this account."""
+
+    @property
+    @abstractmethod
+    def event_account_filter(self):
+        """Get a Django query filter to restrict events to this account."""
+
+    @staticmethod
+    @abstractmethod
+    def get_event_product_identifier(event):
+        """
+        Get the cloud-specific product identifier for an event.
+
+        Args:
+            event (InstanceEvent): The invent to identify
+
+        Returns:
+            str: Some relatively unique string for grouping by product
+
+        """
+
+    @staticmethod
+    @abstractmethod
+    def get_event_instance_identifier(event):
+        """
+        Get the cloud-specific instance identifier for an event.
+
+        Args:
+            event (InstanceEvent): The invent to identify
+
+        Returns:
+            str: Some relatively unique string for grouping by instance
+
+        """
+
+
+class AwsReportHelper(ReportHelper):
+    """Report helper for AWS."""
+
+    def assert_account_exists(self):  # noqa: D102, see parent class
```

Comment:
I don't think there is any harm in inserting an aws specific docstring for this method, better than littering it with `noqa` tags.

Reply:
Yeah, doing this felt iffy, and I was wondering how you guys would feel about it. On the other hand, copy-pasting large swaths of doc strings doesn't feel any better. I wish there was a better supported way to stay DRY with them. 😕

Reply:
Maybe not, but in case we ever use anything to auto generate docs from the methods it would still be nice that there would be something written down.